File: aoc-2023-rust/src/bin/05_01.rs
use aoc::common::day05::{get_mode, ParseMode};
use aoc::parsers::{parse_u64s, split};
use std::io::{self, BufRead};

struct Config {
    src: u64,
    dst: u64,
    distance: u64,
}

fn main() {
    let mut seeds: Vec<u64> = Vec::new();

    let mut seed_to_soil: Vec<Config> = vec![];
    let mut soil_to_fert: Vec<Config> = vec![];
    let mut fert_to_water: Vec<Config> = vec![];
    let mut water_to_light: Vec<Config> = vec![];
    let mut light_to_temp: Vec<Config> = vec![];
    let mut temp_to_hum: Vec<Config> = vec![];
    let mut hum_to_loc: Vec<Config> = vec![];

    let mut mode = ParseMode::Seeds;

    for line_res in io::stdin().lock().lines() {
        let line = line_res.unwrap();

        if line.is_empty() {
            // reached the end of the current config
            // reset mode
            mode = ParseMode::Pending;
            continue;
        }

        match mode {
            ParseMode::Seeds => {
                seeds = parse_seeds(&line);
                mode = ParseMode::Pending;
            }
            ParseMode::SeedToSoil => insert(&mut seed_to_soil, &line),
            ParseMode::SoilToFertilizer => insert(&mut soil_to_fert, &line),
            ParseMode::FertilizerToWater => insert(&mut fert_to_water, &line),
            ParseMode::WaterToLight => insert(&mut water_to_light, &line),
            ParseMode::LightToTemperature => insert(&mut light_to_temp, &line),
            ParseMode::TemperatureToHumidity => insert(&mut temp_to_hum, &line),
            ParseMode::HumidityToLocation => insert(&mut hum_to_loc, &line),
            // determine what should be parsed next
            // empty lines are ignored
            ParseMode::Pending => {
                mode = get_mode(line);
            }
        }
    }

    let mut ans = u64::MAX;
    for seed in seeds {
        let soil = find_next(seed, &seed_to_soil);
        let fert = find_next(soil, &soil_to_fert);
        let water = find_next(fert, &fert_to_water);
        let light = find_next(water, &water_to_light);
        let temp = find_next(light, &light_to_temp);
        let hum = find_next(temp, &temp_to_hum);
        let loc = find_next(hum, &hum_to_loc);

        if loc < ans {
            ans = loc;
        }
    }

    println!("{}", ans);
}

fn insert(src_to_dst: &mut Vec<Config>, line: &str) {
    let config = parse_u64s(&line, " ");
    src_to_dst.push(Config {
        src: config[1],
        dst: config[0],
        distance: config[2],
    });
}

fn parse_seeds(line: &str) -> Vec<u64> {
    parse_u64s(&split(&line, ":")[1], " ")
}

fn find_next(src: u64, src_to_dst: &Vec<Config>) -> u64 {
    let mut needle = src;
    for c in src_to_dst {
        if needle < c.src {
            continue;
        }
        if c.src <= needle && needle <= c.src + c.distance {
            needle = c.dst + (needle - c.src);
            break;
        }
    }
    needle
}

File: aoc-ts/src/main/2023/05_01.ts
import { to1DArrayNumeric } from "../utils/parsers.ts";

export const SEED_TO_SOIL = "seed-to-soil";
export const SOIL_TO_FERTILIZER = "soil-to-fertilizer";
export const FERTILIZER_TO_WATER = "fertilizer-to-water";
export const WATER_TO_LIGHT = "water-to-light";
export const LIGHT_TO_TEMPERATURE = "light-to-temperature";
export const TEMPERATURE_TO_HUMIDITY = "temperature-to-humidity";
export const HUMIDITY_TO_LOCATION = "humidity-to-location";

export function solve(input: string): number {
  const rows = input.trim().split("\n");
  const seeds = to1DArrayNumeric(rows[0].split(":")[1]);
  const map = parseMappings(rows);

  let ans = Infinity;
  for (const seed of seeds) {
    const soil = next(seed, map[SEED_TO_SOIL]);
    const fertilizer = next(soil, map[SOIL_TO_FERTILIZER]);
    const water = next(fertilizer, map[FERTILIZER_TO_WATER]);
    const light = next(water, map[WATER_TO_LIGHT]);
    const temperature = next(light, map[LIGHT_TO_TEMPERATURE]);
    const humidity = next(temperature, map[TEMPERATURE_TO_HUMIDITY]);
    const location = next(humidity, map[HUMIDITY_TO_LOCATION]);
    ans = Math.min(location, ans);
  }

  return ans;
}

function next(n: number, mapping: [number, number, number][]) {
  let m = n;
  for (const [dst, src, rng] of mapping) {
    if (src <= n && n <= src + rng) {
      m = dst + (n - src);
      break;
    }
  }
  return m;
}

export function parseMappings(rows: string[]) {
  const map: { [key: string]: [number, number, number][] } = {
    [SEED_TO_SOIL]: [],
    [SOIL_TO_FERTILIZER]: [],
    [FERTILIZER_TO_WATER]: [],
    [WATER_TO_LIGHT]: [],
    [LIGHT_TO_TEMPERATURE]: [],
    [TEMPERATURE_TO_HUMIDITY]: [],
    [HUMIDITY_TO_LOCATION]: [],
  };

  let i = 0;
  let curr: [number, number, number][] = [];
  while (i < rows.length) {
    if (rows[i].indexOf(SEED_TO_SOIL) >= 0) {
      curr = map[SEED_TO_SOIL];
      i += 1;
    } else if (rows[i].indexOf(SOIL_TO_FERTILIZER) >= 0) {
      curr = map[SOIL_TO_FERTILIZER];
      i += 1;
    } else if (rows[i].indexOf(FERTILIZER_TO_WATER) >= 0) {
      curr = map[FERTILIZER_TO_WATER];
      i += 1;
    } else if (rows[i].indexOf(WATER_TO_LIGHT) >= 0) {
      curr = map[WATER_TO_LIGHT];
      i += 1;
    } else if (rows[i].indexOf(LIGHT_TO_TEMPERATURE) >= 0) {
      curr = map[LIGHT_TO_TEMPERATURE];
      i += 1;
    } else if (rows[i].indexOf(TEMPERATURE_TO_HUMIDITY) >= 0) {
      curr = map[TEMPERATURE_TO_HUMIDITY];
      i += 1;
    } else if (rows[i].indexOf(HUMIDITY_TO_LOCATION) >= 0) {
      curr = map[HUMIDITY_TO_LOCATION];
      i += 1;
    } else if (rows[i].trim().length === 0) {
      curr = [];
      i += 1;
    } else {
      while (i < rows.length && rows[i].trim().length !== 0) {
        curr.push(to1DArrayNumeric(rows[i]) as [number, number, number]);
        i += 1;
      }
      curr.sort((a, b) => a[1] - b[1]);

      // fill in missing intervals
      const missing: [number, number, number][] = [];
      if (curr[0][1] !== 0) {
        missing.push([0, 0, curr[0][1]]);
      }
      let tmp = curr[0][1];
      for (let j = 0; j < curr.length; j++) {
        if (tmp < curr[j][1]) {
          missing.push([tmp, tmp, curr[j][1] - tmp]);
        }
        tmp = curr[j][1] + curr[j][2];
      }
      missing.push([
        curr[curr.length - 1][1] + curr[curr.length - 1][2],
        curr[curr.length - 1][1] + curr[curr.length - 1][2],
        Infinity,
      ]);
      curr.push(...missing);
    }
  }

  return map;
}
